lib/sqlalchemy/dialects/mssql/base.py

```diff
@@ -1872,57 +1873,83 @@ def _opt_encode(self, statement):
         if self.compiled and self.compiled.schema_translate_map:
             rst = self.compiled.preparer._render_schema_translates
             statement = rst(statement, self.compiled.schema_translate_map)
-
         return statement
 
+    def _should_enable_identity_insert(self, id_column, tbl):
+        """Determine whether IDENTITY_INSERT should be enabled."""
+        if self.dialect.warn_on_identity_insert is False:
+            return False
+
+        if id_column is None or isinstance(id_column.default, Sequence):
+            return False
+
+        compile_state = self.compiled.dml_compile_state
+        return id_column.key in self.compiled_parameters[0] or (
+            compile_state._dict_parameters
+            and id_column.key in compile_state._insert_col_keys
+        )
+
+    def _handle_identity_insert_warning(self, tbl):
+        """
+        Handle warnings or errors based on the warn_on_identity_insert flag.
+        """
+        if self.dialect.warn_on_identity_insert == "warn":
+            warnings.warn(
+                f"Automatic identity insert is enabled for table {tbl.name}."
+                "In future versions, this behavior may be disabled by default."
+                "Consider explicitly managing IDENTITY_INSERT.",
+                FutureWarning,
+            )
+        elif not self.dialect.warn_on_identity_insert:
```

Comment:
we can probably just use assert here `assert self.dialect.warn_on_identity_insert`